holoviews/core/data/narwhals.py

```diff
@@ -0,0 +1,620 @@
+    def values(
+        cls,
+        dataset,
+        dim,
+        expanded=True,
+        flat=True,
+        compute=True,
+        keep_index=False,
+    ):
+        dim = dataset.get_dimension(dim, strict=True)
+        data = dataset.data.select(dim.name)
+        is_lazy = isinstance(data, nw.LazyFrame)
+        if not expanded:
+            # LazyFrame does not support maintain_order, it can therefore for some
+            # backends return non-determistic results even for the same data.
+            # It looks like polars.LazyFrame support it, but not duckdb
+            data = data.unique(**({} if is_lazy else {"maintain_order": True}))
+        if is_lazy:
+            if compute:
+                return data.collect()[dim.name]
+            else:
+                return data  # Cannot slice LazyFrame
+        return data[dim.name]
+
+    @classmethod
+    def sample(cls, dataset, samples=None):
+        if samples is None:
+            samples = []
+        data = dataset.data
+        columns = list(data.collect_schema())
+        mask = None
+        for sample in samples:
+            sample_mask = None
+            if np.isscalar(sample):
+                sample = [sample]
+            for col, value in zip(columns, sample, strict=False):
+                submask = nw.col(col) == value
+                if sample_mask is None:
+                    sample_mask = submask
+                else:
+                    sample_mask &= submask
+            if mask is None:
+                mask = sample_mask
+            else:
+                mask |= sample_mask
+        return data.filter(mask)
+
+    @classmethod
+    def add_dimension(cls, dataset, dimension, dim_pos, values, vdim):
+        data = cls._narwhals_clone(dataset.data)
+        cols = list(data.collect_schema())
+        if dimension.name not in cols:
+            cols = [*cols[:dim_pos], dimension.name, *cols[dim_pos:]]
+            if not isinstance(values, nw.Series):
+                if np.isscalar(values):
+                    values = nw.lit(values)
+                else:
+                    values = nw.new_series(
+                        dimension.name,
+                        values,
+                        backend=_EAGER_TYPE.get(data.implementation, data.implementation),
+                    )
+            if isinstance(data, nw.LazyFrame) and isinstance(values, nw.Series):
+                # NOTE(LazyFrame): forced conversion
+                data = data.collect()
+            data = data.with_columns(**{dimension.name: values}).select(cols)
+        return data
+
+    @classmethod
+    def assign(cls, dataset, new_data):
+        return dataset.data.with_columns(**new_data)
+
+    @classmethod
+    def as_dframe(cls, dataset):
+        """Returns the data of a Dataset as a dataframe avoiding copying
+        if it already a dataframe type.
+
+        """
+        if issubclass(dataset.interface, NarwhalsInterface):
+            return dataset.data
```

Comment:
Definitely a considerable broadening of our definition of DataFrame.

Reply:
Do you want to force it to pandas? Or should we try this out for now and enable it if we see too many problems with it?